File: android/app/src/main/java/com/enigo/app/AppState.kt
package com.enigo.app

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.enigo.app.data.Backend
import com.enigo.app.data.ContentData
import com.enigo.app.data.OnboardingProfile
import com.enigo.app.data.OwnProfile
import com.enigo.app.data.ProfilePatch
import com.enigo.app.data.SubscriptionStatus
import com.enigo.app.data.UsernameGenerator
import com.enigo.app.ui.account.LegalDocument
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

sealed class Step {
    data class IntroSlide(val index: Int) : Step()
    object Phone : Step()
    object Verify : Step()
    object Name : Step()
    object Photo : Step()
    object Interests : Step()
    object Bio : Step()
    object Gender : Step()
    object MatchedWith : Step()
    object ShownTo : Step()
    object Community : Step()
    object Location : Step()
    object Intent : Step()
    data class QuestionStep(val index: Int) : Step()
    object NotificationPermission : Step()
    object Searching : Step()
    object MatchReveal : Step()
    object Dashboard : Step()
    data class Chat(val matchId: String) : Step()
    data class Report(val matchId: String) : Step()
    data class SoftExit(val matchId: String) : Step()
    object Profile : Step()
    object Settings : Step()
    object Paywall : Step()
    object Subscription : Step()
    object SignInHelp : Step()
    object DeleteAccountConfirm : Step()
}

/**
 * Root of the vertical slice: one flat state machine over [Step], mirroring
 * ios/Enigo/App/AppState.swift rather than Jetpack Navigation, since the
 * flow is linear and every step feeds the same draft profile.
 */
class AppState : ViewModel() {
    var step by mutableStateOf<Step>(Step.IntroSlide(0)); private set
    var errorMessage by mutableStateOf<String?>(null)
    var isBusy by mutableStateOf(false); private set
    var presentedLegalDocument by mutableStateOf<LegalDocument?>(null)

    var phoneNumber by mutableStateOf("")
    var verifyCode by mutableStateOf("")
    var username by mutableStateOf(UsernameGenerator.generate())
    var firstName by mutableStateOf("")
    var lastName by mutableStateOf("")
    var usernameError by mutableStateOf<String?>(null)
    var photoBytes by mutableStateOf<ByteArray?>(null)
    var selectedInterests by mutableStateOf(setOf<String>())
    var bio by mutableStateOf("")
    var gender by mutableStateOf<String?>(null)
    var genderSelfDescription by mutableStateOf("")
    var matchWith by mutableStateOf(setOf<String>())
    var shownTo by mutableStateOf(setOf<String>())
    var community by mutableStateOf<String?>(null)
    var locationGranted by mutableStateOf(false)
    var lat by mutableStateOf<Double?>(null)
    var lng by mutableStateOf<Double?>(null)
    var radiusKm by mutableStateOf<Int?>(50)
    var intent by mutableStateOf<String?>(null)
    val answers = mutableMapOf<Int, Int>()

    /** The match just found by onboarding's initial search — drives the
     * one-time Match Reveal screen only. */
    var revealedMatchId by mutableStateOf<String?>(null)

    var activeMatchIds by mutableStateOf<List<String>>(emptyList())
    var ownProfile by mutableStateOf<OwnProfile?>(null)
    var ownPhotoURL by mutableStateOf<String?>(null)
    var subscriptionStatus by mutableStateOf<SubscriptionStatus?>(null)
    var rematchCreditsRemaining by mutableStateOf<Int?>(null)
    var rematchUnlimited by mutableStateOf(false)
    var exportedData by mutableStateOf<ByteArray?>(null)

    private var searchJob: Job? = null

    /** Called once at launch. If a session survived and the account already
     * finished onboarding, skip straight to the dashboard. */
    fun bootstrap() = viewModelScope.launch {
        repeat(10) {
            if (Backend.userId.value != null) return@repeat
            delay(200)
        }
        if (Backend.userId.value == null) return@launch
        if (runCatching { Backend.fetchOwnProfile() }.getOrNull() != null) {
            openDashboard()
        }
    }

    fun advanceIntro() {
        val s = step
        step = if (s is Step.IntroSlide && s.index < ContentData.introSlides.size - 1) {
            Step.IntroSlide(s.index + 1)
        } else {
            Step.Phone
        }
    }

    private val normalizedPhone: String get() = phoneNumber.filter { it.isDigit() }

    fun submitPhone() = run {
        Backend.requestOtp(normalizedPhone)
        step = Step.Verify
    }

    fun submitVerify() = run {
        Backend.verifyOtp(normalizedPhone, verifyCode)
        // This phone number may already belong to a completed account (e.g.
        // reinstalling the app, or verifying again after signing out) — in
        // that case skip straight to the dashboard instead of re-running
        // onboarding and overwriting their existing profile.
        if (runCatching { Backend.fetchOwnProfile() }.getOrNull() != null) {
            openDashboard()
        } else {
            step = Step.Name
        }
    }

    /** Mirrors the database trigger (`enforce_username_privacy`): a
     * username may not contain the user's own first or last name, checked
     * case-insensitively and ignoring anything but letters/digits, so the
     * user sees the problem immediately instead of only after the server
     * rejects it. */
    fun validateUsername(): Boolean {
        fun normalize(s: String) = s.lowercase().filter { it.isLetterOrDigit() }
        val normalizedUsername = normalize(username)
        val normalizedFirst = normalize(firstName)
        val normalizedLast = normalize(lastName)
        if (normalizedFirst.isNotEmpty() && normalizedUsername.contains(normalizedFirst)) {
            usernameError = "Username can't contain your first name"
            return false
        }
        if (normalizedLast.isNotEmpty() && normalizedUsername.contains(normalizedLast)) {
            usernameError = "Username can't contain your last name"
            return false
        }
        usernameError = null
        return true
    }

    fun submitName() {
        if (firstName.trim().isEmpty() || lastName.trim().isEmpty() || username.trim().isEmpty()) return
        if (!validateUsername()) return
        step = Step.Photo
    }

    fun submitPhoto() { step = Step.Interests }
    fun submitInterests() { step = Step.Bio }
    fun submitBio() { step = Step.Gender }
    fun submitGender() { step = Step.MatchedWith }
    fun submitMatchedWith() { step = Step.ShownTo }
    fun submitShownTo() { step = Step.Community }
    fun submitCommunity() { step = Step.Location }
    fun submitLocation() { step = Step.Intent }
    fun submitIntent() { step = Step.QuestionStep(0) }

    fun submitAnswer(optionIndex: Int) {
        val s = step
        if (s !is Step.QuestionStep) return
        answers[s.index] = optionIndex
        step = if (s.index + 1 < ContentData.questions.size) {
            Step.QuestionStep(s.index + 1)
        } else {
            Step.NotificationPermission
        }
    }

    /**
     * Uploads the photo (if any), writes the full onboarding profile, and
     * starts searching. This is the single point the draft becomes real —
     * everything before this is only ever held in memory on-device.
     */
    fun completeOnboarding() = run {
        val uid = Backend.userId.value ?: error("Not signed in")
        val photoPath = photoBytes?.let { Backend.uploadPhoto(it, "photo.jpg") }

        val profile = OnboardingProfile(
            id = uid,
            username = username,
            firstName = firstName,
            lastName = lastName,
            gender = gender,
            genderSelfDescription = genderSelfDescription.ifEmpty { null },
            matchWith = matchWith.toList(),
            shownTo = shownTo.toList(),
            community = community,
            intent = intent,
            interests = selectedInterests.toList(),
            bio = bio.ifEmpty { null },
            photoPath = photoPath,
            lat = if (locationGranted) lat else null,
            lng = if (locationGranted) lng else null,
            radiusKm = radiusKm,
            locationGranted = locationGranted,
            answers = answers.entries.associate { (k, v) -> (k + 1).toString() to v },
            onboardingComplete = true
        )
        Backend.upsertProfile(profile)
        step = Step.Searching
        beginSearching()
    }

    /** Free tier holds exactly 1 concurrent match, so an empty dashboard for
     * a free user always means "matchless," not "chose to leave a slot
     * open" — unlike Pro, which can genuinely have spare capacity by
     * choice. Called from DashboardScreen so a free user with nothing
     * active goes straight into the searching flow instead of sitting on an
     * empty "Start a new connection" screen requiring an extra tap. */
    fun autoSearchIfFreeAndEmpty() {
        if (activeMatchIds.isEmpty() && subscriptionStatus?.isPro != true) {
            step = Step.Searching
            beginSearching()
        }
    }

    private fun beginSearching() {
        searchJob?.cancel()
        searchJob = viewModelScope.launch {
            while (true) {
                try {
                    val result = Backend.findMatch()
                    // Cancellation only takes effect at the next suspend
                    // point, so a search that was already superseded by a
                    // fresh beginSearching() call can still resume here
                    // after the fact — re-check before committing its
                    // result to shared state.
                    if (!isActive) return@launch
                    val id = result.matchIds.firstOrNull()
                    if (id != null) {
                        revealedMatchId = id
                        activeMatchIds = result.matchIds
                        step = Step.MatchReveal
                        return@launch
                    }
                } catch (e: CancellationException) {
                    throw e
                } catch (e: Exception) {
                    errorMessage = e.message
                }
                delay(3000)
            }
        }
    }

    fun enterChat() {
        revealedMatchId?.let { step = Step.Chat(it) }
    }

    // MARK: Dashboard

    fun openDashboard() {
        step = Step.Dashboard
        viewModelScope.launch { refreshDashboard() }
    }

    suspend fun refreshDashboard() {
        try {
            activeMatchIds = Backend.listActiveMatchIds()
        } catch (e: Exception) {
            errorMessage = e.message
        }
    }

    fun startNewConnection() = run {
        val result = Backend.findMatch()
        activeMatchIds = result.matchIds
    }

    fun openChat(matchId: String) { step = Step.Chat(matchId) }
    fun openReport(matchId: String) { step = Step.Report(matchId) }
    fun openSoftExit(matchId: String) { step = Step.SoftExit(matchId) }

    fun submitReport(matchId: String, category: String, detail: String?) = run {
        Backend.submitReport(matchId, category, detail)
        refreshDashboard()
        step = Step.Dashboard
    }

    fun confirmSoftExit(matchId: String, reason: String?) = run {
        val result = Backend.endMatch(matchId, reason)
        rematchCreditsRemaining = result.rematchCreditsRemaining
        rematchUnlimited = result.unlimited
        refreshDashboard()
        step = Step.Dashboard
    }

    // MARK: Profile / settings

    suspend fun loadOwnProfile() {
        try {
            ownProfile = Backend.fetchOwnProfile()
            ownProfile?.photoPath?.let { path ->
                ownPhotoURL = runCatching { Backend.ownPhotoURL(path) }.getOrNull()
            }
        } catch (e: Exception) {
            errorMessage = e.message
        }
    }

    fun patchProfile(patch: ProfilePatch) = run {
        Backend.patchProfile(patch)
        loadOwnProfile()
    }

    fun openProfile() {
        step = Step.Profile
        viewModelScope.launch { loadOwnProfile() }
    }

    fun openSettings() {
        step = Step.Settings
        viewModelScope.launch { loadOwnProfile() }
    }

    fun retakeQuestions() {
        answers.clear()
        step = Step.QuestionStep(0)
    }

    fun openSignInHelp() { step = Step.SignInHelp }
    fun openDeleteAccountConfirm() { step = Step.DeleteAccountConfirm }

    // MARK: Pro / subscription

    suspend fun loadSubscriptionStatus() {
        try {
            subscriptionStatus = Backend.getSubscriptionStatus()
        } catch (e: Exception) {
            errorMessage = e.message
        }
    }

    fun openPaywall() {
        step = Step.Paywall
        viewModelScope.launch { loadSubscriptionStatus() }
    }

    fun openSubscription() {
        step = Step.Subscription
        viewModelScope.launch { loadSubscriptionStatus() }
    }

    fun recordPurchase(productId: String, transactionId: String, currentPeriodEndIso: String?) = run {
        Backend.recordPurchase(productId, transactionId, currentPeriodEndIso)
        loadSubscriptionStatus()
        step = Step.Dashboard
    }

    fun redeemBoost(productId: String, transactionId: String) = run {
        Backend.redeemBoost(productId, transactionId)
    }

    // MARK: Account

    suspend fun exportAccountData() {
        try {
            exportedData = Backend.exportAccountData()
        } catch (e: Exception) {
            errorMessage = e.message
        }
    }

    fun deleteAccount() = run {
        Backend.deleteAccount()
        resetToIntro()
    }

    fun signOut() = run {
        Backend.signOut()
        resetToIntro()
    }

    private fun resetToIntro() {
        searchJob?.cancel()
        phoneNumber = ""
        verifyCode = ""
        photoBytes = null
        selectedInterests = emptySet()
        bio = ""
        gender = null
        matchWith = emptySet()
        shownTo = emptySet()
        community = null
        locationGranted = false
        lat = null
        lng = null
        intent = null
        answers.clear()
        revealedMatchId = null
        activeMatchIds = emptyList()
        ownProfile = null
        subscriptionStatus = null
        step = Step.IntroSlide(0)
    }

    private fun run(block: suspend () -> Unit) {
        viewModelScope.launch {
            isBusy = true
            errorMessage = null
            try {
                block()
            } catch (e: Exception) {
                errorMessage = e.message ?: "Something went wrong"
            }
            isBusy = false
        }
    }
}

File: ios/Enigo/App/AppState.swift
import CoreLocation
import Foundation
import Supabase
import SwiftUI

enum Step: Equatable {
    case ageVerification
    case introSlide(Int)
    case phone
    case verify
    case name
    case photo
    case interests
    case bio
    case gender
    case matchedWith
    case shownTo
    case community
    case location
    case intent
    case question(Int)
    case notificationPermission
    case searching
    case matchReveal
    case dashboard
    case chat(UUID)
    case report(UUID)
    case softExit(UUID)
    case profile
    case settings
    case paywall
    case subscription
    case signInHelp
    case deleteAccountConfirm
}

/// Root of the vertical slice: one flat state machine over `Step`, mirroring
/// the prototype's single-component `state.screen` (see Enigo.dc.html) rather
/// than a NavigationStack, since the flow is linear and every step feeds the
/// same draft profile.
@MainActor
final class AppState: ObservableObject {
    @Published var step: Step = .ageVerification
    @Published var errorMessage: String?
    @Published var isBusy = false
    @Published var presentedLegalDocument: LegalDocument?

    // Onboarding draft — persisted to the backend in one upsert once
    // notification permission is resolved (see `completeOnboarding`).
    @Published var birthdate = Date()
    @Published var phoneNumber = ""
    @Published var verifyCode = ""
    @Published var username = UsernameGenerator.generate()
    @Published var firstName = ""
    @Published var lastName = ""
    @Published var usernameError: String?
    @Published var photoData: Data?
    var photoFileName: String?
    @Published var selectedInterests: Set<String> = []
    @Published var bio = ""
    @Published var gender: String?
    @Published var genderSelfDescription = ""
    @Published var matchWith: Set<String> = []
    @Published var shownTo: Set<String> = []
    @Published var community: String?
    @Published var locationGranted = false
    @Published var locationCoordinate: CLLocationCoordinate2D?
    @Published var radiusKm: Int? = 50
    @Published var intent: String?
    @Published var answers: [Int: Int] = [:]

    private let locationManager = LocationManager()

    /// Requests real device location when the user flips "Share my rough
    /// location" on. Reverts the toggle if permission is denied or no fix
    /// could be obtained — the toggle only ever reflects a fix we actually
    /// have, never an aspirational state.
    func requestLocationPermission() async {
        if let coordinate = await locationManager.requestOneShotLocation() {
            locationCoordinate = coordinate
            locationGranted = true
        } else {
            locationCoordinate = nil
            locationGranted = false
        }
    }

    /// The match just found by onboarding's initial search — used only to
    /// drive the one-time Match Reveal screen.
    @Published var revealedMatchId: UUID?

    // MARK: - Dashboard / account state

    @Published var activeMatchIds: [UUID] = []
    @Published var ownProfile: OwnProfile?
    @Published var ownPhotoURL: URL?
    @Published var subscriptionStatus: SubscriptionStatus?

    private let backend = Backend.shared
    private var searchTask: Task<Void, Never>?

    /// Called once at launch. If a session survived in the Keychain and the
    /// account already finished onboarding, skip straight to the dashboard
    /// instead of dumping a returning user back into onboarding.
    func bootstrap() async {
        for _ in 0..<10 {
            if backend.userId != nil { break }
            try? await Task.sleep(for: .milliseconds(200))
        }
        guard backend.userId != nil else { return }
        if (try? await backend.fetchOwnProfile()) != nil {
            openDashboard()
        }
    }

    // MARK: - Onboarding navigation

    func advanceIntro() {
        if case .introSlide(let i) = step, i < ContentData.introSlides.count - 1 {
            step = .introSlide(i + 1)
        } else {
            step = .phone
        }
    }

    /// Digits only, no leading '+' — matches how the phone-auth test_otp
    /// numbers are configured locally (see backend/supabase/config.toml).
    private var normalizedPhone: String {
        phoneNumber.filter(\.isNumber)
    }

    func submitPhone() async {
        await run {
            try await self.backend.requestOTP(phone: self.normalizedPhone)
            self.step = .verify
        }
    }

    func submitVerify() async {
        await run {
            try await self.backend.verifyOTP(phone: self.normalizedPhone, code: self.verifyCode)
            // This phone number may already belong to a completed account
            // (e.g. reinstalling the app, or verifying again after signing
            // out) — in that case skip straight to the dashboard instead of
            // re-running onboarding and overwriting their existing profile.
            if (try? await self.backend.fetchOwnProfile()) != nil {
                self.openDashboard()
            } else {
                self.step = .name
            }
        }
    }

    /// True if `username` (normalized: lowercased, letters/digits only) contains
    /// `firstName` or `lastName` the same way normalized — mirrors the
    /// database trigger (`enforce_username_privacy`) so the user sees the
    /// problem immediately instead of only after the server rejects it.
    func validateUsername() -> Bool {
        func normalize(_ s: String) -> String {
            String(s.lowercased().filter { $0.isLetter || $0.isNumber })
        }
        let normalizedUsername = normalize(username)
        let normalizedFirst = normalize(firstName)
        let normalizedLast = normalize(lastName)
        if !normalizedFirst.isEmpty && normalizedUsername.contains(normalizedFirst) {
            usernameError = "Username can't contain your first name"
            return false
        }
        if !normalizedLast.isEmpty && normalizedUsername.contains(normalizedLast) {
            usernameError = "Username can't contain your last name"
            return false
        }
        usernameError = nil
        return true
    }

    func submitName() {
        guard !firstName.trimmingCharacters(in: .whitespaces).isEmpty,
              !lastName.trimmingCharacters(in: .whitespaces).isEmpty,
              !username.trimmingCharacters(in: .whitespaces).isEmpty,
              validateUsername()
        else { return }
        step = .photo
    }

    func submitPhoto() { step = .interests }
    func submitInterests() { step = .bio }
    func submitBio() { step = .gender }
    func submitGender() { step = .matchedWith }
    func submitMatchedWith() { step = .shownTo }
    func submitShownTo() { step = .community }
    func submitCommunity() { step = .location }
    func submitLocation() { step = .intent }
    func submitIntent() { step = .question(0) }

    func submitAnswer(_ optionIndex: Int) {
        guard case .question(let i) = step else { return }
        answers[i] = optionIndex
        if i + 1 < ContentData.questions.count {
            withAnimation(EnigoMotion.tapAdvance) { step = .question(i + 1) }
        } else {
            step = .notificationPermission
        }
    }

    /// Uploads the photo (if any), writes the full onboarding profile, and
    /// starts searching. This is the single point the draft becomes real —
    /// everything before this is only ever held in memory on-device.
    func completeOnboarding() async {
        await run {
            guard let userId = self.backend.userId else { throw BackendError.notSignedIn }

            var photoPath: String?
            if let data = self.photoData {
                photoPath = try await self.backend.uploadPhoto(data: data, fileName: self.photoFileName ?? "photo.jpg")
            }

            let profile = OnboardingProfile(
                id: userId,
                username: self.username,
                firstName: self.firstName,
                lastName: self.lastName,
                birthdate: self.birthdate,
                gender: self.gender,
                genderSelfDescription: self.genderSelfDescription.isEmpty ? nil : self.genderSelfDescription,
                matchWith: Array(self.matchWith),
                shownTo: Array(self.shownTo),
                community: self.community,
                intent: self.intent,
                interests: Array(self.selectedInterests),
                bio: self.bio.isEmpty ? nil : self.bio,
                photoPath: photoPath,
                lat: self.locationGranted ? self.locationCoordinate?.latitude : nil,
                lng: self.locationGranted ? self.locationCoordinate?.longitude : nil,
                radiusKm: self.radiusKm,
                locationGranted: self.locationGranted,
                answers: Dictionary(uniqueKeysWithValues: self.answers.map { (String($0.key + 1), $0.value) }),
                onboardingComplete: true
            )
            try await self.backend.upsertProfile(profile)
            self.step = .searching
            self.beginSearching()
        }
    }

    // MARK: - Matching

    /// Free tier holds exactly 1 concurrent match, so an empty dashboard for
    /// a free user always means "matchless," not "chose to leave a slot
    /// open" — unlike Pro, which can genuinely have spare capacity by
    /// choice. Called from DashboardView so a free user with nothing active
    /// goes straight into the searching flow instead of sitting on an empty
    /// "Start a new connection" screen requiring an extra tap.
    func autoSearchIfFreeAndEmpty() {
        guard activeMatchIds.isEmpty, subscriptionStatus?.isPro != true else { return }
        step = .searching
        beginSearching()
    }

    private func beginSearching() {
        searchTask?.cancel()
        searchTask = Task {
            while !Task.isCancelled {
                do {
                    let result = try await backend.findMatch()
                    // Cancelling searchTask doesn't abort this in-flight
                    // call (Swift's task cancellation is cooperative), so a
                    // search that was already superseded by a fresh
                    // beginSearching() can still resolve here after the
                    // fact — re-check before committing its result.
                    guard !Task.isCancelled else { return }
                    if let id = result.matchIds.first {
                        revealedMatchId = id
                        activeMatchIds = result.matchIds
                        step = .matchReveal
                        return
                    }
                } catch {
                    errorMessage = error.localizedDescription
                }
                try? await Task.sleep(for: .seconds(3))
            }
        }
    }

    func enterChat() {
        guard let id = revealedMatchId else { return }
        step = .chat(id)
    }

    // MARK: - Dashboard

    func openDashboard() {
        step = .dashboard
        Task { await refreshDashboard() }
    }

    func refreshDashboard() async {
        do {
            activeMatchIds = try await backend.listActiveMatchIds()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    /// "Start a new connection" — searches for one more slot (free tier is a
    /// no-op once its single slot is full; Pro can hold up to 3).
    func startNewConnection() async {
        await run {
            let result = try await self.backend.findMatch()
            self.activeMatchIds = result.matchIds
        }
    }

    func openChat(_ matchId: UUID) { step = .chat(matchId) }
    func openReport(_ matchId: UUID) { step = .report(matchId) }
    func openSoftExit(_ matchId: UUID) { step = .softExit(matchId) }

    func submitReport(matchId: UUID, category: String, detail: String?) async {
        await run {
            try await self.backend.submitReport(matchId: matchId, category: category, detail: detail)
            await self.refreshDashboard()
            self.step = .dashboard
        }
    }

    @Published var rematchCreditsRemaining: Int?
    @Published var rematchUnlimited = false

    func confirmSoftExit(matchId: UUID, reason: String?) async {
        await run {
            let result = try await self.backend.endMatch(matchId: matchId, reason: reason)
            self.rematchCreditsRemaining = result.rematchCreditsRemaining
            self.rematchUnlimited = result.unlimited
            await self.refreshDashboard()
            self.step = .dashboard
        }
    }

    // MARK: - Profile / settings

    func loadOwnProfile() async {
        do {
            ownProfile = try await backend.fetchOwnProfile()
            if let path = ownProfile?.photoPath {
                ownPhotoURL = try? await backend.ownPhotoURL(path: path)
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func patchProfile(_ patch: ProfilePatch) async {
        await run {
            try await self.backend.patchProfile(patch)
            await self.loadOwnProfile()
        }
    }

    func openProfile() {
        step = .profile
        Task { await loadOwnProfile() }
    }

    func openSettings() {
        step = .settings
        Task { await loadOwnProfile() }
    }

    // MARK: - Pro / subscription

    func loadSubscriptionStatus() async {
        do {
            subscriptionStatus = try await backend.getSubscriptionStatus()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func openPaywall() {
        step = .paywall
        Task { await loadSubscriptionStatus() }
    }

    func openSubscription() {
        step = .subscription
        Task { await loadSubscriptionStatus() }
    }

    func recordPurchase(productId: String, transactionId: String, expiresAt: Date?) async {
        await run {
            try await self.backend.recordPurchase(productId: productId, transactionId: transactionId, currentPeriodEnd: expiresAt)
            await self.loadSubscriptionStatus()
            self.step = .dashboard
        }
    }

    func redeemBoost(productId: String, transactionId: String) async {
        await run {
            try await self.backend.redeemBoost(productId: productId, transactionId: transactionId)
        }
    }

    // MARK: - Account

    @Published var exportedData: Data?

    func exportAccountData() async {
        await run {
            self.exportedData = try await self.backend.exportAccountData()
        }
    }

    func deleteAccount() async {
        await run {
            try await self.backend.deleteAccount()
            self.resetToIntro()
        }
    }

    func signOut() async {
        await run {
            try await self.backend.signOut()
            self.resetToIntro()
        }
    }

    private func resetToIntro() {
        searchTask?.cancel()
        phoneNumber = ""
        verifyCode = ""
        photoData = nil
        selectedInterests = []
        bio = ""
        gender = nil
        matchWith = []
        shownTo = []
        community = nil
        locationGranted = false
        locationCoordinate = nil
        intent = nil
        answers = [:]
        revealedMatchId = nil
        activeMatchIds = []
        ownProfile = nil
        subscriptionStatus = nil
        step = .introSlide(0)
    }

    // MARK: - Helpers

    private func run(_ body: @escaping () async throws -> Void) async {
        isBusy = true
        errorMessage = nil
        do {
            try await body()
        } catch let error as PostgrestError {
            errorMessage = error.message
        } catch {
            errorMessage = Backend.friendlyMessage(from: error)
        }
        isBusy = false
    }
}
